derp/xdp/xdp.c

```diff
@@ -0,0 +1,585 @@
+	} else {
+		sa->length = bpf_htons(sizeof(struct stunxor));
+	}
+
+	struct stunxor *xor;
+	struct stunxor6 *xor6;
+
+	// Adjust tail and reset header pointers.
+	int adjust_tail_by;
+	if (is_ipv6) {
+		xor6 = attr_data;
+		adjust_tail_by = (void *)(xor6 + 1) - data_end;
+	} else {
+		xor = attr_data;
+		adjust_tail_by = (void *)(xor + 1) - data_end;
+	}
+	if (bpf_xdp_adjust_tail(ctx, adjust_tail_by)) {
+		return XDP_ABORTED;
+	}
+	data_end = (void *)(long)ctx->data_end;
+	data = (void *)(long)ctx->data;
+	eth = data;
+	if ((void *)(eth + 1) > data_end) {
+		return XDP_ABORTED;
+	}
+	if (is_ipv6) {
+		ip6 = (void *)(eth + 1);
+		if ((void *)(ip6 + 1) > data_end) {
+			return XDP_ABORTED;
+		}
+		udp = (void *)(ip6 + 1);
+		if ((void *)(udp + 1) > data_end) {
+			return XDP_ABORTED;
+		}
+	} else {
+		ip = (void *)(eth + 1);
+		if ((void *)(ip + 1) > data_end) {
+			return XDP_ABORTED;
+		}
+		udp = (void *)(ip + 1);
+		if ((void *)(udp + 1) > data_end) {
+			return XDP_ABORTED;
+		}
+	}
+	req = (void *)(udp + 1);
+	if ((void *)(req + 1) > data_end) {
+		return XDP_ABORTED;
+	}
+	sa = (void *)(req + 1);
+	if ((void *)(sa + 1) > data_end) {
+		return XDP_ABORTED;
+	}
+
+	// Set attr data.
+	if (is_ipv6) {
+		xor6 = (void *)(sa + 1);
+		if ((void *)(xor6 + 1) > data_end) {
+			return XDP_ABORTED;
+		}
+		xor6->unused = 0x00; // unused byte
+		xor6->family = 0x02;
+		xor6->port = udp->source ^ bpf_htons(STUN_MAGIC_FOR_PORT_XOR);
+		xor6->addr[0] = ip6->saddr.in6_u.u6_addr32[0] ^ bpf_htonl(STUN_MAGIC);
+		for (int i = 1; i < 4; i++) {
+			// All three are __be32, no endianness flips.
+			xor6->addr[i] = ip6->saddr.in6_u.u6_addr32[i] ^ req->txid[i-1];
+		}
+	} else {
+		xor = (void *)(sa + 1);
+		if ((void *)(xor + 1) > data_end) {
+			return XDP_ABORTED;
+		}
+		xor->unused = 0x00; // unused byte
+		xor->family = 0x01;
+		xor->port = udp->source ^ bpf_htons(STUN_MAGIC_FOR_PORT_XOR);
+		xor->addr = ip->saddr ^ bpf_htonl(STUN_MAGIC);
+	}
+
+	// Flip ethernet header source and destination address.
+	__u8 eth_tmp[ETH_ALEN];
+	__builtin_memcpy(eth_tmp, eth->h_source, ETH_ALEN);
+	__builtin_memcpy(eth->h_source, eth->h_dest, ETH_ALEN);
+	__builtin_memcpy(eth->h_dest, eth_tmp, ETH_ALEN);
+
+	// Flip ip header source and destination address.
+	if (is_ipv6) {
+		struct in6_addr ip_tmp = ip6->saddr;
+		ip6->saddr = ip6->daddr;
+		ip6->daddr = ip_tmp;
+	} else {
+		__be32 ip_tmp = ip->saddr;
+		ip->saddr = ip->daddr;
+		ip->daddr = ip_tmp;
+	}
+
+	// Flip udp header source and destination ports;
+	__be16 port_tmp = udp->source;
+	udp->source = udp->dest;
+	udp->dest = port_tmp;
+
+	// Update total length, TTL, and checksum.
+	__u32 cs = 0;
+	if (is_ipv6) {
+		if ((void *)(ip6 +1) > data_end) {
+			return XDP_ABORTED;
+		}
+		__u16 payload_len = data_end - (void *)(ip6 + 1);
+		ip6->payload_len = bpf_htons(payload_len);
+		ip6->hop_limit = 64;
+	} else {
+		__u16 tot_len = data_end - (void *)ip;
+		ip->tot_len = bpf_htons(tot_len);
+		ip->ttl = 64;
```

Comment:
optional, minor: `IPDEFTTL` maybe?

Reply:
updated in https://github.com/tailscale/tailscale/pull/12390/commits/abb689d12db1b17b4b43f1057e54ce614fe6a7a1